Add new health check to service

nshumoogum opened the pull request:
### What

Add new health endpoint with the following checkers:

* elasticsearch
* hierarchy api
* kafka consumer
* kafka producer
* kafka error producer (for reporter)

Re-worked entry point to service via main and renamed service package to event and moved graceful shutdown logic to run functionality to be consistent with other applications.

Also includes:
* Use new go dependency manage, modules instead of vendor
* Use latest libraries, dp-kafka, dp-rchttp, dp-api-clients-go, log.go, dp-elasticsearch

### How to review

* Check unit tests pass
* Run an integration test
* Check code changes make sense

### Who can review

Anyone

nshumoogum committed:
Update dependency manager

Use go modules instead of go vendor

nshumoogum committed:
Merge branch 'develop' into feature/health-check

nshumoogum committed:
Update ci scripts

nshumoogum committed:
Update service to handle new health check

Also includes reworking main function and service package to event package

nshumoogum committed:
Update nomad plan to include http service health check

nshumoogum committed:
Use latest healthcheck library

Rename variables and add comment to exported variable

nshumoogum committed:
Update nomad plan, allow more time for application to turn healthy (green) in environment

CarlHembrough commented on config/config.go:
```diff
@@ -8,21 +8,21 @@ import (
+	EventReporterTopic         string        `envconfig:"EVENT_REPORTER_TOPIC"`
+	GracefulShutdownTimeout    time.Duration `envconfig:"GRACEFUL_SHUTDOWN_TIMEOUT"`
+	HealthCheckInterval        time.Duration `envconfig:"HEALTHCHECK_INTERVAL"`
+	HealthCheckCriticalTimeout time.Duration `envconfig:"HEALTHCHECK_CRITICAL_TIMEOUT"`
```
This also needs to be updated in the readme

CarlHembrough commented on event/handler.go:
```diff
@@ -85,7 +91,7 @@ func (svc *Service) handleMessage(ctx context.Context, message kafka.Message) (s
 
 		if err = apis.addChildrenToSearchIndex(ctx, instanceID, dimension, codeID); err != nil {
-			log.Error(err, log.Data{"instance_id": instanceID, "dimension": dimension, "code_id": codeID})
+			log.Event(ctx, "faile to add children dimension options", log.ERROR, log.Error(err), log.Data{"instance_id": instanceID, "dimension": dimension, "code_id": codeID})
```
typo "faile"

CarlHembrough commented on go.mod:
```diff
@@ -0,0 +1,39 @@
+	github.com/ONSdigital/dp-hierarchy-api v0.0.0-20200309110613-76f2517bcbed
+	github.com/ONSdigital/dp-import v0.0.0-20180202121531-d3cc28e452c3
+	github.com/ONSdigital/dp-kafka v1.1.5
+	github.com/ONSdigital/dp-rchttp v0.0.0-20200114090501-463a529590e8
```
Can this be replaced with a release version?

nshumoogum replied:
There is currently no release for this library and James was updating the library with headers - https://github.com/ONSdigital/dp-rchttp/pull/9 - not sure whether we should be creating release 1 or not.

nshumoogum replied:
Actually I will create the first release and update the dependency here

CarlHembrough commented on main.go:
```diff
@@ -2,99 +2,265 @@ package main
+		// If kafka consumer exists, stop listening to it. (Will close later)
+		if serviceList.Consumer {
+			log.Event(shutdownContext, "stopping kafka consumer listener", log.INFO)
+			syncConsumerGroup.StopListeningToConsumer(shutdownContext)
```
Should we be handling the error return value here?

CarlHembrough commented on main.go:
```diff
@@ -2,99 +2,265 @@ package main
+		// If search builder error kafka producer exists, close it
+		if serviceList.SearchBuilderErrProducer {
+			log.Event(shutdownContext, "closing search builder error kafka producer", log.INFO, log.Data{"topic": cfg.EventReporterTopic})
+			searchBuilderErrProducer.Close(shutdownContext)
```
Should we be handling the error return value here?

CarlHembrough commented on main.go:
```diff
@@ -2,99 +2,265 @@ package main
+		// If search built kafka producer exists, close it
+		if serviceList.SearchBuiltProducer {
+			log.Event(shutdownContext, "closing search builder kafka producer", log.INFO, log.Data{"topic": cfg.ProducerTopic})
+			searchBuiltProducer.Close(shutdownContext)
```
Should we be handling the error return value here?

CarlHembrough commented on main.go:
```diff
@@ -2,99 +2,265 @@ package main
+		// If kafka consumer exists, close it
+		if serviceList.Consumer {
+			log.Event(shutdownContext, "closing kafka consumer", log.INFO, log.Data{"consumer": "SyncConsumerGroup"})
+			syncConsumerGroup.Close(shutdownContext)
```
Should we be handling the error return value here?

CarlHembrough commented on main.go:
```diff
@@ -2,99 +2,265 @@ package main
+
+	if err = hc.AddCheck("Kafka Search Built Producer", searchBuiltProducer.Checker); err != nil {
+		hasErrors = true
+		log.Event(ctx, "error adding check for kafka import producer", log.ERROR, log.Error(err))
```
"error adding check for kafka search built producer"?

CarlHembrough reviewed (changes requested):
Good work - just a few minor points to consider

nshumoogum committed:
Update dp-rchttp dependency

Use semver release tag v1.0.0

nshumoogum committed:
Add missing application config documentation to README.md

nshumoogum committed:
Fix typo

Also update README.md

nshumoogum committed:
Handle errors returned from closing kafka connections

Also update logging messages to make logs easier to read (consistent) when gracefully shutting down

nshumoogum committed:
Refactor handling of shutdown errors
